src/EditorFeatures/CSharp/CommentSelection/CSharpToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CommentSelection;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Editor.Implementation.CommentSelection;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Operations;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+using VSCommanding = Microsoft.VisualStudio.Commanding;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.CommentSelection
+{
+    [Export(typeof(VSCommanding.ICommandHandler))]
+    [ContentType(ContentTypeNames.CSharpContentType)]
+    [Name(PredefinedCommandHandlerNames.ToggleBlockComment)]
+    internal class CSharpToggleBlockCommentCommandHandler :
+        ToggleBlockCommentCommandHandler
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        internal CSharpToggleBlockCommentCommandHandler(
+            ITextUndoHistoryRegistry undoHistoryRegistry,
+            IEditorOperationsFactoryService editorOperationsFactoryService,
+            ITextStructureNavigatorSelectorService navigatorSelectorService)
+            : base(undoHistoryRegistry, editorOperationsFactoryService, navigatorSelectorService)
+        {
+        }
+
+        /// <summary>
+        /// Retrieves block comments in document using the CSharp syntax tree.
+        /// </summary>
+        protected override async Task<ImmutableArray<TextSpan>> GetBlockCommentsInDocument(Document document, ITextSnapshot snapshot,
+            CommentSelectionInfo commentInfo, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
this might too expansive since it make you to iterate whole tree and then keep only ones you want. you probably want to give in the span to DescendantTrivia
http://source.roslyn.io/#Microsoft.CodeAnalysis/Syntax/SyntaxNode.cs,1035

Reply:
span is overlapwith so probably work for partial overalp span case as well.

Reply:
There is a case where I need to get non-overlapping comments on the same line.  I guess I could pass in a span from the start of the first line in the selections until the end of the last line in the selections.

I'll try it out to see if it works.